### PR TITLE
[icmporch] use query attribute capability for stats mode and set explicitly supported value

### PR DESCRIPTION
**Pull request by ayurkiv-nvda**
Depend on PR: https://github.com/sonic-net/sonic-sairedis/pull/1756 
It should fix PR failure

**What I did**
use query attribute capability for stats mode and set an explicitly supported value.
If stats mode is not set or not supported then log WARN messages and rely on internal SAI logic during session creation

**Why I did it**
need to check if some SAI capabilities are supported before using them.
Currently there are few possible stats count mode:
- SAI_STATS_COUNT_MODE_PACKET_AND_BYTE,
- SAI_STATS_COUNT_MODE_PACKET
- SAI_STATS_COUNT_MODE_BYTE
- SAI_STATS_COUNT_MODE_NONE

Preffered mode: 
1) PACKET_AND_BYTE,  2) PACKET only,  3)BYTE only,  4) none

**How I verified it**
UT test added 

**Details if related**

### CODE REVIEW

**Review by prsunny — DISMISSED** [review]
lgtm, @ayurkiv-nvda , please update the PR description with more details and the details on which attribute being queried.

also, please check the build failure

**Comment by prsunny** [comment]
@zjswhhh , please review

**Comment by ayurkiv-nvda** [comment]
Hello @prsunny 
This https://github.com/sonic-net/sonic-sairedis/pull/1756 should fix VS failure in PR checks.
But I see that PR checks are still failing with the same error, which means that [1756 ](https://github.com/sonic-net/sonic-sairedis/pull/1756) is not yet available in [docker-sonic-vs.gz](https://dev.azure.com/mssonic/build/_build/results?buildId=1089827&view=logs&j=3f6395b2-1619-5ebe-f305-2aedcf353cb5&t=49e6bbf0-3bec-5f80-2da5-a76c50ca6929), which is used for PR checks

Can someone please check if we are using the latest docker-sonic-vs.gz with needed fix on a stage "Download pre-stage built docker-sonic-vs"?

**Review by zjswhhh — DISMISSED** [review]

**Review by volodymyrsamotiy — DISMISSED** [review]

**Comment by ayurkiv-nvda** [comment]
Coverage test failed; going to update PR with UT

**Review by volodymyrsamotiy — DISMISSED** [review]

**Comment by volodymyrsamotiy** [comment]
@prsunny, could you please re-approve and help to merge? Now PR checks passed.

**Comment by ayurkiv-nvda** [comment]
Hello @prsunny @zjswhhh 
I have added UT test increase coverage, now PR checks passed successfully
Can you please help re-approve and help to merge?

**Comment by liat-grozovik** [comment]
have few questions/suggestion

1. sai_metadata_get_attr_metadata() and sai_query_attribute_enum_values_capability() both query hardware capabilities that i assume do not change at runtime. Calling them inside do_create() means every ICMP session creation pays this cost and re-does the same query. Should it be moved so init part instead of on each create?
2. should u consider adding test to cover the below as well? test PACKET_AND_BYTE and NONE-only rejection
3. is FAILED_VALID_ENTRY should be used when the capabilities are not supported?

File: orchagent/icmporch.cpp [hunk 1]
```diff
@@ -559,6 +560,65 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
         m_fv_map[m_tx_interval_fname] = "0";
     }
 
+    // Query capability for stats count mode and set a supported value explicitly
+    {
+        const auto* meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_ICMP_ECHO_SESSION,
+                                                          SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE);
+        if (!meta || !meta->isenum)
+        {
+            SWSS_LOG_ERROR("sai_metadata_get_attr_metadata for SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE failed ");
```
Comment (prsunny) [inline orchagent/icmporch.cpp]: Why do we need ERROR here, it can be a WARN, right?

**Review by zjswhhh — DISMISSED** [review]

File: orchagent/icmporch.cpp [hunk 2]
```diff
@@ -559,6 +560,65 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
...
+
+        auto status = sai_query_attribute_enum_values_capability(
+            gSwitchId,
+            SAI_OBJECT_TYPE_ICMP_ECHO_SESSION,
+            SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE,
+            &values);
+
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("sai_query_attribute_enum_values_capability for SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE failed ");
```
Comment (prsunny) [inline orchagent/icmporch.cpp]: Same here

**Comment by zjswhhh** [comment]
@manamand2020 - please review as well.

File: orchagent/icmporch.cpp [hunk 3]
```diff
@@ -559,6 +560,65 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
...
+
+        if (set)
+        {
+            m_attr_val_map[SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE] = val;
+            m_fv_vector.push_back({m_stats_count_mode_fname, "PACKET"});
+        }
+        else
+        {
+            SWSS_LOG_ERROR("No supported stats count mode found");
+            return SaiOffloadHandlerStatus::FAILED_VALID_ENTRY;
```
Comment (zjswhhh) [inline orchagent/icmporch.cpp]: Should we fail the creation completely?

File: orchagent/icmporch.cpp [hunk 4]
```diff
@@ -331,6 +331,7 @@ const std::string IcmpSaiSessionHandler::m_hw_lookup_fname          = "hw_lookup
 const std::string IcmpSaiSessionHandler::m_nexthop_switchover_fname = "nexthop_switchover";
 const std::string IcmpSaiSessionHandler::m_session_type_normal      = "NORMAL";
 const std::string IcmpSaiSessionHandler::m_session_type_rx          = "RX";
+const std::string IcmpSaiSessionHandler::m_stats_count_mode_fname   = "stats_count_mode";
```
Comment (manamand2020) [inline orchagent/icmporch.cpp]: Do we need this per session?

File: orchagent/icmporch.cpp [hunk 5]
```diff
@@ -559,6 +560,65 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
         m_fv_map[m_tx_interval_fname] = "0";
     }
 
+    // Query capability for stats count mode and set a supported value explicitly
```
Comment (manamand2020) [inline orchagent/icmporch.cpp]: Do we need this query per session?

File: orchagent/icmporch.cpp [hunk 6]
```diff
@@ -559,6 +560,65 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
...
+            {
+                val.s32 = SAI_STATS_COUNT_MODE_PACKET;
+                set = true;
+            }
+        }
+
+        if (set)
+        {
+            m_attr_val_map[SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE] = val;
+            m_fv_vector.push_back({m_stats_count_mode_fname, "PACKET"});
```
Comment (manamand2020) [inline orchagent/icmporch.cpp]: Do you need 2 values here based val used "PACKET" and "PACKET_N_BYTE"?

File: orchagent/icmporch.cpp (hunk 3 above)
Comment (manamand2020) [reply orchagent/icmporch.cpp]: We should not tie this to session creation and this should be a global capability of icmporch.

**Review by manamand2020 — COMMENTED** [review]
Can we check the capability once during initialization of icmporch and use the same COUNT_MODE for all sessions.

Can we can relax the stats capability and allow session creation?

File: orchagent/icmporch.cpp (hunk 3 above)
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: Done

File: orchagent/icmporch.cpp (hunk 5 above)
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: Done

File: orchagent/icmporch.cpp (hunk 4 above)
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: Done, now it is global

File: orchagent/icmporch.cpp (hunk 6 above)
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: Fixed, now we are using direct mapping in naming (PACKET_AND_BYTE, PACKET, BYTE, NONE)

File: orchagent/icmporch.cpp (hunk 1 above)
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: I double checked SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE , it is not marked as mandatory parameter, so I agree, it can be WARN

File: orchagent/icmporch.cpp [hunk 7]
```diff
@@ -559,6 +631,19 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
         m_fv_map[m_tx_interval_fname] = "0";
     }
 
+    if (m_orch.m_stats_count_mode_initialized)
+    {
+        sai_attribute_value_t statsCountModeAttr{};
+        statsCountModeAttr.s32 = m_orch.m_stats_count_mode;
+        m_attr_val_map[SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE] = statsCountModeAttr;
+        m_fv_vector.push_back({m_stats_count_mode_fname, m_orch.m_stats_count_mode_str});
```
Comment (manamand2020) [inline orchagent/icmporch.cpp]: Why do you need this in redis-db? As this is global setting, so why do we need to set this in redis-db for each session?

File: orchagent/icmporch.h [hunk 8]
```diff
@@ -164,6 +175,12 @@ class IcmpOrch: public Orch, public Subject
     // keeps track of number of sessions
     uint32_t m_num_sessions = 0;
 
+    // resolved ICMP stats count mode reused across session creates
+    sai_stats_count_mode_t m_stats_count_mode = SAI_STATS_COUNT_MODE_PACKET;
+    std::string m_stats_count_mode_str = "PACKET";
+    bool m_stats_count_mode_initialized = false;
+
+
```
Comment (manamand2020) [inline orchagent/icmporch.h]: remove this extra line

File: orchagent/icmporch.cpp (hunk 7 above)
Comment (prsunny) [reply orchagent/icmporch.cpp]: +1 on this. Can we write only one time to DB, As I understand, mode would be same for all sessions

**Comment by prsunny** [comment]
Overall, lgtm,  @ayurkiv-nvda could you address copilot comments and change log levels? @volodymyrsamotiy for viz

File: orchagent/icmporch.cpp [hunk 9]
```diff
@@ -73,6 +78,72 @@ IcmpOrch::~IcmpOrch(void)
...
+        m_stats_count_mode = SAI_STATS_COUNT_MODE_PACKET_AND_BYTE;
+        m_stats_count_mode_str = "PACKET_AND_BYTE";
+        m_stats_count_mode_initialized = true;
+    }
+    else if (std::find(values.list, end, static_cast<int32_t>(SAI_STATS_COUNT_MODE_PACKET)) != end)
+    {
+        m_stats_count_mode = SAI_STATS_COUNT_MODE_PACKET;
+        m_stats_count_mode_str = "PACKET";
+        m_stats_count_mode_initialized = true;
+    }
```
[the remark this answers is not in the document]
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp]: Overall logic changed now: use a list of predefined modes and check one-by-one from the highest prio to the lowest

File: orchagent/icmporch.cpp [hunk 10]
```diff
@@ -73,6 +78,72 @@ IcmpOrch::~IcmpOrch(void)
...
+
+    const auto *meta = sai_metadata_get_attr_metadata(
+        SAI_OBJECT_TYPE_ICMP_ECHO_SESSION,
+        SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE);
+    if (!meta || !meta->isenum)
+    {
+        SWSS_LOG_ERROR("sai_metadata_get_attr_metadata for SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE failed");
+        return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (ayurkiv-nvda) [reply orchagent/icmporch.cpp:94]: Changed loglevel

**Comment by ayurkiv-nvda** [comment]
Hello @prsunny , @manamand2020 
All comments are fixed, preffered order omplemenation improved
PR checks passed
Can we approve and merge PR?

**Review by manamand2020 — APPROVED** [review]
LGTM

File: orchagent/icmporch.cpp [hunk 11]
```diff
@@ -559,6 +617,17 @@ SaiOffloadHandlerStatus IcmpSaiSessionHandler::do_create()
...
 
+    if (m_orch.m_stats_count_mode_initialized)
+    {
+        sai_attribute_value_t statsCountModeAttr{};
+        statsCountModeAttr.s32 = m_orch.m_stats_count_mode;
+        m_attr_val_map[SAI_ICMP_ECHO_SESSION_ATTR_STATS_COUNT_MODE] = statsCountModeAttr;
+    }
+    else
+    {
+        SWSS_LOG_WARN("%s, Stats count mode capability unresolved", m_name.c_str());
```
Comment (manamand2020) [inline orchagent/icmporch.cpp:628]: Remove this warning as this was emitted during the IcmpOrch construction phase.

**Review by prsunny — APPROVED** [review]
